akka-projection-core/src/main/scala/akka/projection/Projection.scala

```diff
@@ -134,4 +134,22 @@ private[projection] trait RunningProjectionManagement[Offset] {
   def setOffset(offset: Option[Offset]): Future[Done]
   def getManagementState(): Future[Option[ManagementState]]
   def setPaused(paused: Boolean): Future[Done]
+
+  /** Not necessarily available for projections with all sources: future will fail with IllegalStateException
+   *  if source does not support this operation
+   */
+  def getSourceMaxOffset(): Future[Option[Offset]]
+
+  /** Computes the lag between the last saved offset for the projection and the max offset emittable by the
+   *  source.
+   *
+   *  The returned future should fail with IllegalStateException if getSourceMaxOffset() fails with an
+   *  IllegalStateException.
+   *
+   *  The returned future will complete with None if lag could not be computed now, but might be computable later.
+   *  The returned future will complete positive if the saved offset is behind the source max offset.
+   *  The returned future will complete zero if exactly caught up.
+   *  The returned future will complete non-positive if the saved offset is ahead of max offset.
+   */
+  def getLag(): Future[Option[Long]]
```

Comment:
While we're here, I wonder if we should have both the max offset and the count of events since current offset. So we can pass an offset (which for timestamp offset includes both the timestamp and the seen seq numbers for that offset) and have a full backlog status, with both how many events behind we are and the time difference.

Reply:
Could add that... implementations wouldn't be required to support that, but would avoid the possibility (e.g. with a pure time-based lag, there could be an initial spike after a period of inactivity (e.g. no events over a weekend and then Monday morning the first event comes in and if the lag check for publishing metrics happens before the commit, that would show 60 hours or something of lag).

Reply:
One scheme that might be good for deciding whether to scale out projections could be to use this one (which could be by convention a faster check) first and then if over some threshold do an event count (which would likely be a heavier operation, whether in the DB or in processing).